### ISSUE TITLE
[Problem]: crashes when client had shoddy wifi signal

### ISSUE THREAD

**Issue by dasl-** [mikebrady/shairport-sync#1565]
### What happened?

Hi there, 

Today I was playing from an iOS spotify client to four SPS receivers running on raspberry pis. I walked to a corner of my apartment that has worse wifi reception. I think my phone may have momentarily lost the wifi signal. When this happened, two of the four SPS receivers crashed. The crashes were different.

One SPS receiver crashed with SIGABRT, and one crashed with SIGSEGV.

# SIGABRT logs
Here are the last 10,000 log lines of the SPS receiver that crashed with SIGABRT: https://gist.github.com/dasl-/90c932c536536b0c032ff9eb99ea21ab

Probably the most relevant bit is at the end:
```
Oct 29 04:42:13 piwall10 shairport-sync:          0.000136389 "player.c:1084" Accepting packet 48203 with timestamp 274592293. Lead time is 1.993424 seconds.
Oct 29 04:42:13 piwall10 shairport-sync[4416]: shairport-sync: malloc.c:4043: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.
Oct 29 04:42:13 piwall10 shairport-sync:          0.000047722 "player.c:2410" Player: packets out of sequence: expected: 48194, got: 48203, with ab_read: 48204 and ab_write: 48213.
Oct 29 04:42:13 piwall10 systemd[1]: shairport-sync.service: Main process exited, code=killed, status=6/ABRT
Oct 29 04:42:13 piwall10 systemd[1]: shairport-sync.service: Failed with result 'signal'.
```

Here is a gdb backtrace of the coredump from this crash: https://gist.github.com/dasl-/f0d6b8db1227bb3e5868ed88319ab545

# SIGSEGV logs
Here are the last 10,000 log lines of the SPS receiver that crashed with SIGSEGV: https://gist.github.com/dasl-/2f9102bfe579d6f7a3165f9f12c7684c

Probably the most relevant bit is at the end:
```
Oct 29 04:42:12 kitchen shairport-sync:          0.000015537 "player.c:1088" Packet 45264 empty or not ready.
Oct 29 04:42:13 kitchen systemd[1]: shairport-sync.service: Main process exited, code=killed, status=11/SEGV
Oct 29 04:42:13 kitchen systemd[1]: shairport-sync.service: Failed with result 'signal'.
Oct 29 04:42:13 kitchen systemd[1]: shairport-sync.service: Consumed 27min 43.813s CPU time.
```

Here is a gdb backtrace of the coredump from this crash: https://gist.github.com/dasl-/1ff8bcd8ab87392a06449128f2294c6f

Let me know if you'd like me to run any other gdb commands.

### Relevant log output

```shell
See above
```

### Configuration Information.

both receivers had the same configuration:

```
[piwall10] ~ % shairport-sync --displayConfig
         0.000812685 "shairport.c:1420" default metadata_pipename is "/tmp/shairport-sync-metadata".
         0.000122278 "shairport.c:2057" >> Display Config Start.
         0.010242481 "shairport.c:2057"
         0.000067889 "shairport.c:2057" From "uname -a":
         0.000018926 "shairport.c:2057"  Linux piwall10 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l GNU/Linux
         0.014765111 "shairport.c:2057"
         0.000077945 "shairport.c:2057" From /etc/os-release:
         0.000019574 "shairport.c:2057"  Raspbian GNU/Linux 10 (buster)
         0.011954555 "shairport.c:2057"
         0.000067926 "shairport.c:2057" From /sys/firmware/devicetree/base/model:
         0.000045833 "shairport.c:2057"  Raspberry Pi 4 Model B Rev 1.2
         0.000098149 "shairport.c:2057"
         0.000046111 "shairport.c:2057" Shairport Sync Version String:
         0.000040407 "shairport.c:2057"  4.1-rc0-162-gfca26490-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
         0.000043093 "shairport.c:2057"
         0.000038926 "shairport.c:2057" Command Line:
         0.000039111 "shairport.c:2057"  shairport-sync --displayConfig
         0.000095444 "shairport.c:2057"
         0.000043963 "shairport.c:2057" Configuration File:
         0.000043334 "shairport.c:2057"  /etc/shairport-sync.conf
         0.000040148 "shairport.c:2057"
         0.000222000 "shairport.c:2057" Configuration File Settings:
         0.000048074 "shairport.c:2057"  general :
         0.000040315 "shairport.c:2057"  {
         0.000039740 "shairport.c:2057"    volume_range_db = 40;
         0.000040463 "shairport.c:2057"    volume_max_db = 0.0;
         0.000042926 "shairport.c:2057"    name = "piwall";
         0.000040056 "shairport.c:2057"    audio_backend_buffer_desired_length_in_seconds = 2;
         0.000041222 "shairport.c:2057"  };
         0.000039741 "shairport.c:2057"  alsa :
         0.000043000 "shairport.c:2057"  {
         0.000043259 "shairport.c:2057"    output_device = "hw:Headphones";
         0.000043722 "shairport.c:2057"    mixer_control_name = "Headphone";
         0.000034093 "shairport.c:2057"  };
         0.000015759 "shairport.c:2057"  diagnostics :
         0.000014111 "shairport.c:2057"  {
         0.000118833 "shairport.c:2057"    log_verbosity = 2;
         0.000020371 "shairport.c:2057"    statistics = "yes";
         0.000016241 "shairport.c:2057"  };
         0.000015963 "shairport.c:2057"
         0.000015388 "shairport.c:2057" >> Display Config End.
         0.000016519 "shairport.c:2059" >> Goodbye!
         0.000032518 "shairport.c:1534" exit function called...
         0.000017186 "shairport.c:1552" Stopping D-Bus service
         0.000051370 "dbus-service.c:1131" stopping dbus service
         0.000044982 "shairport.c:1554" Stopping D-Bus service done
         0.000039981 "shairport.c:1570" Stopping DACP Monitor
         0.000043056 "shairport.c:1572" Stopping DACP Monitor Done
         0.000043111 "shairport.c:1576" Stopping metadata hub
         0.000043092 "shairport.c:1578" Stopping metadata done
         0.000042574 "shairport.c:1582" Stopping metadata
         0.000060130 "shairport.c:1584" Stopping metadata done
         0.000045222 "shairport.c:1586" Stopping the activity monitor.
         0.000040000 "shairport.c:1588" Stopping the activity monitor done.
         0.000059093 "shairport.c:1670" normal exit
```

### How did you install Shairport Sync?

Built from source

### Check previous issues

- [X] Confirm

**Comment by mikebrady** [comment]
Wow, excellent stuff, thanks. Let me dig through this. Can I ask what the very long `audio_backend_buffer_desired_length_in_seconds` is for, please?

**Comment by mikebrady** [comment]
Thanks again for this. So, a couple of things.

1. The crashes seems to be caused by memory allocation / management errors.
2. I am unable to find that commit in the `git` repository.
3. I wonder about that very long `audio_backend_buffer_desired_length_in_seconds`, whether it could be causing any problems.

So, for (3), would it be possible to, say, comment out the `audio_backend_buffer_desired_length_in_seconds` setting and see if the crashes still occur, please? If they do, for (2), it would be great if you could delete or totally put aside the existing Shairport Sync directory you are building out of, pull a fresh version from GitHub into a completely separate directory, build it and try it out.

**Comment by dasl-** [comment]
>I wonder about that very long audio_backend_buffer_desired_length_in_seconds, whether it could be causing any problems.

I was attempting to fix a newly occurring issue I'm experiencing: brief 1-3 second audio drop outs a handful of times per hour as per these troubleshooting instructions: https://github.com/mikebrady/shairport-sync/blob/development/TROUBLESHOOTING.md#buffer-underflow-to-audio-backend

I was going to try to collect some more data before bothering you with a github issue about it 😅 , but maybe I'll post some info here since you're curious. I'm doing several new things recently, so I'm not sure yet which factor is the cause of these newly occurring audio drop outs:

1. Upgraded client from iOS 15 to iOS 16.1
2. Upgraded versions of SPS
3. making more use of multiroom audio features (now playing to 4 SPS receviers at the same time, whereas the previous max I had was 2).
4. using apps (spotify) that generate realtime audio streams rather than buffered audio.

I'll continue to try to collect more data over time though and possibly eventually file a standalone github issue.

Here are the full logs of a run where there were several drop outs (warning, 85 megabyte file 😅 ): https://drive.google.com/file/d/1qtjDXFADYzjrSlIADNmbS7DWbriZfXHH/view?usp=share_link 

In case it's helpful, here is a much more manageable subset of the above logs, filtered for the `recovering from a previous underrun` lines and the statistics log lines: https://gist.github.com/dasl-/bc3a71b4cc8115cfc211a8daebfa3caf

I am a bit confused by these drop outs because unlike previous times I've had drop outs, the network indicators (`Missing |   Late | Too Late | Resend Reqs`) are all `0`.

Anyway, feel free to ignore this lengthy aside until I have more data.

When I get some time (probably next week), I'll work on trying to reproduce these crashes with and without the long `audio_backend_buffer_desired_length_in_seconds`.

**Comment by dasl-** [comment]
Today I tried reproducing these crashes, using the same `audio_backend_buffer_desired_length_in_seconds = 2` value that I had in these two crash instances. I tried walking outside to get out of wifi range over 10 times, while connected to 4 SPS receivers in my apartment. Unfortunately I was not able to reproduce any of these crashes. So it seems like it may be a tough bug to reproduce, possibly subject to some race conditions or similar. I'm going to close this for now since I am unable to reproduce.
